docs: qemu board runner table shows debug even when only west debugserver is supported

### Describe the bug

The board-level "Programming and Debugging" table in the docs is confusing for QEMU boards.

For example, the page shows a table with a `debug` column for a QEMU board (shown in the screenshot for `qemu_cortex_m3`). I initially interpreted this as meaning I could run:

- `west debug`

But in practice, QEMU only supports:

- `west debugserver`

and not `west debug` (at least for this case).

This makes the table misleading because it suggests `west debug` support when the practical workflow is `debugserver` only.

<img width="943" height="235" alt="Image" src="https://github.com/user-attachments/assets/5f3aa766-2730-4a5c-bd9a-12e8d9e1c1e8" />

### Regression

- [ ] This is a regression.

### Steps to reproduce

1. Open the board documentation page that includes the "Programming and Debugging" table (example: `qemu_cortex_m3` shown in screenshot).
2. See that the table includes a `debug` column.
3. Try `west debug` for that board.
4. Observe that the board supports `west debugserver` workflow, not `west debug`.

### Relevant log output

```shell
❯ git describe 
v4.3.0-6701-g28793caf8272
❯ rm -rf build && west build -b qemu_cortex_m3 -t debug zephyr/samples/hello_world          
-- west build: generating a build system
Loading Zephyr default modules (Zephyr base).
-- Application: /home/yashi/src/zephyr-project/zephyr/samples/hello_world
-- CMake version: 4.2.3
-- Found Python3: /home/yashi/src/zephyr-project/.venv/bin/python (found suitable version "3.13.12", minimum required is "3.10") found components: Interpreter
-- Cache files will be written to: /home/yashi/.cache/zephyr
-- Zephyr version: 4.3.99 (/home/yashi/src/zephyr-project/zephyr)
-- Found west (found suitable version "1.5.0", minimum required is "0.14.0")
-- Board: qemu_cortex_m3, qualifiers: ti_lm3s6965
-- ZEPHYR_TOOLCHAIN_VARIANT not set, trying to locate Zephyr SDK
-- Found host-tools: zephyr 0.17.4 (/home/yashi/zephyr-sdk-0.17.4)
-- Found toolchain: zephyr 0.17.4 (/home/yashi/zephyr-sdk-0.17.4)
-- Found Dtc: /home/yashi/zephyr-sdk-0.17.4/sysroots/x86_64-pokysdk-linux/usr/bin/dtc (found suitable version "1.7.0", minimum required is "1.4.6")
-- Found BOARD.dts: /home/yashi/src/zephyr-project/zephyr/boards/qemu/cortex_m3/qemu_cortex_m3.dts
-- Generated zephyr.dts: /home/yashi/src/zephyr-project/build/zephyr/zephyr.dts
-- Generated pickled edt: /home/yashi/src/zephyr-project/build/zephyr/edt.pickle
-- Generated devicetree_generated.h: /home/yashi/src/zephyr-project/build/zephyr/include/generated/zephyr/devicetree_generated.h
Parsing /home/yashi/src/zephyr-project/zephyr/Kconfig
Loaded configuration '/home/yashi/src/zephyr-project/zephyr/boards/qemu/cortex_m3/qemu_cortex_m3_defconfig'
Merged configuration '/home/yashi/src/zephyr-project/zephyr/samples/hello_world/prj.conf'
Configuration saved to '/home/yashi/src/zephyr-project/build/zephyr/.config'
Kconfig header saved to '/home/yashi/src/zephyr-project/build/zephyr/include/generated/zephyr/autoconf.h'
-- Found GnuLd: /home/yashi/zephyr-sdk-0.17.4/arm-zephyr-eabi/arm-zephyr-eabi/bin/ld.bfd (found version "2.38")
-- The C compiler identification is GNU 12.2.0
-- The CXX compiler identification is GNU 12.2.0
-- The ASM compiler identification is GNU
-- Found assembler: /home/yashi/zephyr-sdk-0.17.4/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc
-- Using ccache: /usr/bin/ccache
-- Found gen_kobject_list: /home/yashi/src/zephyr-project/zephyr/scripts/build/gen_kobject_list.py
-- Configuring done (4.0s)
-- Generating done (0.1s)
-- Build files have been written to: /home/yashi/src/zephyr-project/build
-- west build: running target debug
[0/1] Debugging qemu_cortex_m3
WARNING: CMake debug target is deprecated, call west directly instead
-- west debug: rebuilding
[1/127] Preparing syscall dependency handling

[3/127] Generating include/generated/zephyr/version.h
-- Zephyr version: 4.3.99 (/home/yashi/src/zephyr-project/zephyr), build: v4.3.0-6701-g28793caf8272
[127/127] Linking C executable zephyr/zephyr.elf
Memory region         Used Size  Region Size  %age Used
           FLASH:        9440 B       256 KB      3.60%
             RAM:        4104 B        64 KB      6.26%
        IDT_LIST:          0 GB        32 KB      0.00%
Generating files from /home/yashi/src/zephyr-project/build/zephyr/zephyr.elf for board: qemu_cortex_m3/ti_lm3s6965
FATAL ERROR: no runners.yaml found in /home/yashi/src/zephyr-project/build/zephyr. Either board qemu_cortex_m3/ti_lm3s6965 doesn't support west flash/debug/simulate, or a pristine build is needed.
FAILED: [code=1] zephyr/cmake/flash/CMakeFiles/debug /home/yashi/src/zephyr-project/build/zephyr/cmake/flash/CMakeFiles/debug 
cd /home/yashi/src/zephyr-project/build && cmake -E echo WARNING:\ CMake\ debug\ target\ is\ deprecated,\ call\ west\ directly\ instead && /usr/bin/cmake -DTARGET=debug -DDEPENDENCIES="" -P /home/yashi/src/zephyr-project/zephyr/cmake/flash/check_runner_dependencies.cmake && /usr/bin/cmake -E env ZEPHYR_BASE=/home/yashi/src/zephyr-project/zephyr /home/yashi/src/zephyr-project/.venv/bin/python -m west debug
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: /usr/bin/cmake --build /home/yashi/src/zephyr-project/build --target debug
```

### Impact

Annoyance – Minor irritation; no significant impact on usability or functionality.

### Environment

_No response_

### Additional Context

I suspect the table may be generated from runner metadata (possibly `board_runner_info`), but I am not sure how the command/support matrix is currently derived from `doc/_extensions/`.

If the table is auto-generated, it may need to distinguish between:

- `debug`
- `debugserver`
- `debugserver_<emu>`

instead of collapsing them in a way that implies broader support than exists.

--- a/.ruff-excludes.toml
+++ b/.ruff-excludes.toml
@@ -1062,7 +1062,6 @@ exclude = [
     "./scripts/west_commands/runners/openocd.py",
     "./scripts/west_commands/runners/probe_rs.py",
     "./scripts/west_commands/runners/pyocd.py",
-    "./scripts/west_commands/runners/qemu.py",
     "./scripts/west_commands/runners/renode-robot.py",
     "./scripts/west_commands/runners/renode.py",
     "./scripts/west_commands/runners/silabs_commander.py",

--- a/boards/arm/mps2/board.cmake
+++ b/boards/arm/mps2/board.cmake
@@ -63,8 +63,6 @@ elseif(CONFIG_BOARD_MPS2_AN521_CPU0 OR CONFIG_BOARD_MPS2_AN521_CPU0_NS OR CONFIG
   endif()
 endif()
 
-board_set_debugger_ifnset(qemu)
-
 set(ARMFVP_FLAGS ${ARMFVP_FLAGS}
 -C fvp_mps2.telnetterminal0.start_telnet=0
 -C fvp_mps2.telnetterminal1.start_telnet=0
@@ -77,3 +75,5 @@ set(ARMFVP_FLAGS ${ARMFVP_FLAGS}
 -C fvp_mps2.UART2.unbuffered_output=1
 -C fvp_mps2.mps2_visualisation.disable-visualisation=1
 )
+
+include(${ZEPHYR_BASE}/boards/common/qemu.board.cmake)

--- a/boards/arm/mps3/board.cmake
+++ b/boards/arm/mps3/board.cmake
@@ -53,7 +53,7 @@ elseif(CONFIG_BOARD_MPS3_CORSTONE310_FVP OR CONFIG_BOARD_MPS3_CORSTONE310_FVP_NS
   endif()
 endif()
 
-board_set_debugger_ifnset(qemu)
+include(${ZEPHYR_BASE}/boards/common/qemu.board.cmake)
 
 if(CONFIG_BUILD_WITH_TFM)
   # Override the binary used by qemu, to use the combined

--- a/boards/common/qemu.board.cmake
+++ b/boards/common/qemu.board.cmake
@@ -1,0 +1,5 @@
+# SPDX-License-Identifier: Apache-2.0
+
+board_set_flasher_ifnset(qemu)
+board_set_debugger_ifnset(qemu)
+board_finalize_runner_args(qemu)

--- a/boards/qemu/arc/board.cmake
+++ b/boards/qemu/arc/board.cmake
@@ -41,3 +41,4 @@ list(APPEND QEMU_FLAGS_${ARCH}
   )
 
 set(BOARD_DEBUG_RUNNER qemu)
+include(${ZEPHYR_BASE}/boards/common/qemu.board.cmake)

--- a/boards/qemu/cortex_a53/board.cmake
+++ b/boards/qemu/cortex_a53/board.cmake
@@ -32,4 +32,4 @@ if(CONFIG_XIP)
   )
 endif()
 
-board_set_debugger_ifnset(qemu)
+include(${ZEPHYR_BASE}/boards/common/qemu.board.cmake)

--- a/boards/qemu/cortex_a9/board.cmake
+++ b/boards/qemu/cortex_a9/board.cmake
@@ -17,4 +17,4 @@ set(QEMU_KERNEL_OPTION
   "-device;loader,file=\$<TARGET_FILE:\${logical_target_for_zephyr_elf}>,cpu-num=0"
   )
 
-board_set_debugger_ifnset(qemu)
+include(${ZEPHYR_BASE}/boards/common/qemu.board.cmake)

--- a/boards/qemu/cortex_m0/board.cmake
+++ b/boards/qemu/cortex_m0/board.cmake
@@ -10,4 +10,4 @@ set(QEMU_FLAGS_${ARCH}
   -cpu ${QEMU_CPU_TYPE_${ARCH}}
   -machine microbit
   )
-board_set_debugger_ifnset(qemu)
+include(${ZEPHYR_BASE}/boards/common/qemu.board.cmake)

--- a/boards/qemu/cortex_m3/board.cmake
+++ b/boards/qemu/cortex_m3/board.cmake
@@ -8,4 +8,4 @@ set(QEMU_FLAGS_${ARCH}
   -cpu ${QEMU_CPU_TYPE_${ARCH}}
   -machine lm3s6965evb
   )
-board_set_debugger_ifnset(qemu)
+include(${ZEPHYR_BASE}/boards/common/qemu.board.cmake)

--- a/boards/qemu/cortex_r5/board.cmake
+++ b/boards/qemu/cortex_r5/board.cmake
@@ -17,4 +17,4 @@ set(QEMU_KERNEL_OPTION
   "-device;loader,addr=0xff9a0000,data=0x80000218,data-len=4"
   )
 
-board_set_debugger_ifnset(qemu)
+include(${ZEPHYR_BASE}/boards/common/qemu.board.cmake)

--- a/boards/qemu/kvm_arm64/board.cmake
+++ b/boards/qemu/kvm_arm64/board.cmake
@@ -25,4 +25,4 @@ if(CONFIG_XIP)
   )
 endif()
 
-board_set_debugger_ifnset(qemu)
+include(${ZEPHYR_BASE}/boards/common/qemu.board.cmake)

--- a/boards/qemu/leon3/board.cmake
+++ b/boards/qemu/leon3/board.cmake
@@ -10,4 +10,4 @@ set(QEMU_FLAGS_${ARCH}
   -m 1G
   -icount auto
   )
-board_set_debugger_ifnset(qemu)
+include(${ZEPHYR_BASE}/boards/common/qemu.board.cmake)

--- a/boards/qemu/malta/board.cmake
+++ b/boards/qemu/malta/board.cmake
@@ -9,4 +9,4 @@ set(QEMU_FLAGS_${ARCH}
   -serial null
   -serial null
   )
-board_set_debugger_ifnset(qemu)
+include(${ZEPHYR_BASE}/boards/common/qemu.board.cmake)

--- a/boards/qemu/or1k/board.cmake
+++ b/boards/qemu/or1k/board.cmake
@@ -10,4 +10,4 @@ set(QEMU_FLAGS_${ARCH}
   -nographic
 )
 
-board_set_debugger_ifnset(qemu)
+include(${ZEPHYR_BASE}/boards/common/qemu.board.cmake)

--- a/boards/qemu/riscv32/board.cmake
+++ b/boards/qemu/riscv32/board.cmake
@@ -34,4 +34,4 @@ set(QEMU_FLAGS_${ARCH}
   -cpu ${qemu_riscv_cpu}
   )
 
-board_set_debugger_ifnset(qemu)
+include(${ZEPHYR_BASE}/boards/common/qemu.board.cmake)

--- a/boards/qemu/riscv32_xip/board.cmake
+++ b/boards/qemu/riscv32_xip/board.cmake
@@ -8,4 +8,4 @@ set(QEMU_FLAGS_${ARCH}
   -machine sifive_e
 )
 
-board_set_debugger_ifnset(qemu)
+include(${ZEPHYR_BASE}/boards/common/qemu.board.cmake)

--- a/boards/qemu/riscv32e/board.cmake
+++ b/boards/qemu/riscv32e/board.cmake
@@ -34,4 +34,4 @@ set(QEMU_FLAGS_${ARCH}
   -cpu ${qemu_riscv_cpu}
   )
 
-board_set_debugger_ifnset(qemu)
+include(${ZEPHYR_BASE}/boards/common/qemu.board.cmake)

--- a/boards/qemu/riscv64/board.cmake
+++ b/boards/qemu/riscv64/board.cmake
@@ -34,4 +34,4 @@ set(QEMU_FLAGS_${ARCH}
   -cpu ${qemu_riscv_cpu}
   )
 
-board_set_debugger_ifnset(qemu)
+include(${ZEPHYR_BASE}/boards/common/qemu.board.cmake)

--- a/boards/qemu/rx/board.cmake
+++ b/boards/qemu/rx/board.cmake
@@ -14,4 +14,4 @@ if(CONFIG_XIP)
   )
 endif()
 
-board_set_debugger_ifnset(qemu)
+include(${ZEPHYR_BASE}/boards/common/qemu.board.cmake)

--- a/boards/qemu/x86/board.cmake
+++ b/boards/qemu/x86/board.cmake
@@ -2,7 +2,6 @@
 # Copyright (c) 2019 Intel Corp.
 
 set(SUPPORTED_EMU_PLATFORMS qemu)
-
 if(NOT CONFIG_REBOOT)
   set(REBOOT_FLAG -no-reboot)
 endif()
@@ -78,11 +77,6 @@ if(NOT CONFIG_ACPI)
   list(APPEND QEMU_FLAGS_${ARCH} -machine acpi=off)
 endif()
 
-# TODO: Support debug
-# board_set_debugger_ifnset(qemu)
-# debugserver: QEMU_EXTRA_FLAGS += -s -S
-# debugserver: qemu
-
 if(CONFIG_BOARD_QEMU_X86_TINY AND CONFIG_DEMAND_PAGING
    AND NOT CONFIG_LINKER_GENERIC_SECTIONS_PRESENT_AT_BOOT)
   # This is to map the flash so it is accessible.
@@ -90,3 +84,4 @@ if(CONFIG_BOARD_QEMU_X86_TINY AND CONFIG_DEMAND_PAGING
   set(X86_EXTRA_GEN_MMU_ARGUMENTS
       --map ${CONFIG_FLASH_BASE_ADDRESS},${QEMU_FLASH_SIZE_KB},W)
 endif()
+include(${ZEPHYR_BASE}/boards/common/qemu.board.cmake)

--- a/boards/qemu/xtensa/board.cmake
+++ b/boards/qemu/xtensa/board.cmake
@@ -10,7 +10,4 @@ if(CONFIG_BOARD_QEMU_XTENSA)
   )
 endif()
 
-# TODO: Support debug
-# board_set_debugger_ifnset(qemu)
-# debugserver: QEMU_EXTRA_FLAGS += -s -S
-# debugserver: qemu
+include(${ZEPHYR_BASE}/boards/common/qemu.board.cmake)

--- a/scripts/west_commands/runners/qemu.py
+++ b/scripts/west_commands/runners/qemu.py
@@ -2,13 +2,25 @@
 #
 # SPDX-License-Identifier: Apache-2.0
 
-'''Runner stub for QEMU.'''
+'''Runner for QEMU.'''
+
+import subprocess
 
 from runners.core import RunnerCaps, ZephyrBinaryRunner
 
+DEFAULT_GDB_PORT = 1234
+
 
 class QemuBinaryRunner(ZephyrBinaryRunner):
-    '''Place-holder for QEMU runner customizations.'''
+    '''Emulator runner for QEMU.'''
+
+    def __init__(self, cfg, tui=False, gdb_port=DEFAULT_GDB_PORT):
+        super().__init__(cfg)
+        if cfg.gdb is None:
+            self.gdb_cmd = None
+        else:
+            self.gdb_cmd = [cfg.gdb] + (['-tui'] if tui else [])
+        self.gdb_port = gdb_port
 
     @classmethod
     def name(cls):
@@ -16,16 +28,62 @@ class QemuBinaryRunner(ZephyrBinaryRunner):
 
     @classmethod
     def capabilities(cls):
-        # This is a stub.
-        return RunnerCaps(commands=set())
+        return RunnerCaps(commands={'flash', 'debug', 'debugserver'})
 
     @classmethod
     def do_add_parser(cls, parser):
-        pass                    # Nothing to do.
+        pass  # Nothing to do.
 
     @classmethod
     def do_create(cls, cfg, args):
         return QemuBinaryRunner(cfg)
 
     def do_run(self, command, **kwargs):
-        pass
+        # For QEMU emulation, we rely on the CMake targets where available
+        # (run_qemu, debugserver_qemu, etc.) to actually run QEMU. This runner
+        # is installed to allow west debug/debugserver commands to work with
+        # QEMU.
+        #
+        # The actual QEMU invocation happens via the CMake build system.
+        # Here we trigger the appropriate CMake target automatically.
+        if command == 'debugserver':
+            self.debugserver(**kwargs)
+        elif command == 'debug':
+            self.debug(**kwargs)
+        elif command == 'flash':
+            self.do_flash(**kwargs)
+
+    def do_flash(self, **kwargs):
+        """Start QEMU using existing target."""
+        # Trigger the CMake debugserver_qemu target
+        self._run_cmake_target('run_qemu')
+
+    def debug(self, **kwargs):
+        if self.gdb_cmd is None:
+            raise ValueError('Cannot debug; gdb is missing')
+
+        gdb_cmd = self.gdb_cmd + ['-ex', f'target remote :{self.gdb_port}', self.cfg.elf_file]
+        self.require(gdb_cmd[0])
+
+        self.run_client(gdb_cmd)
+
+    def debugserver(self, **kwargs):
+        """Start QEMU with GDB server enabled."""
+        # Trigger the CMake debugserver_qemu target
+        self._run_cmake_target('debugserver_qemu')
+
+    def _run_cmake_target(self, target):
+        """Run a CMake target in the build directory."""
+        build_dir = self.cfg.build_dir
+
+        # Run cmake --build with the target
+        cmd = ['cmake', '--build', str(build_dir), '-t', target]
+
+        self.logger.info(f'Starting QEMU with target: {target}')
+        self.logger.debug(f'Running: {" ".join(cmd)}')
+
+        try:
+            subprocess.run(cmd, check=True)
+        except subprocess.CalledProcessError as e:
+            self.logger.error(f'Failed to run {target}: {e}')
+            raise

